src/esp/gfx/Renderer.cpp

```diff
@@ -86,39 +41,9 @@ struct Renderer::Impl {
 
     draw(sceneGraph.getDefaultRenderCamera(), sceneGraph.getDrawables());
   }
-
-  void readFrameRgba(uint8_t* ptr) {
-    framebuffer_.mapForRead(GL::Framebuffer::ColorAttachment{0});
-    Image2D rgbaImage =
```

Comment:
It would be cool if Image2D, or something like it, could wrap an EGLImage. You could then keep the pixels in GPU memory and only read into CPU memory if the user asked for a CPU pointer. Similar to CVPixelBuffer's CVPixelBufferLockBaseAddress.

https://developer.apple.com/documentation/corevideo/cvpixelbuffer
https://www.khronos.org/registry/EGL/sdk/docs/man/html/eglCreateImage.xhtml

Reply:
There's [GL::BufferImage](https://doc.magnum.graphics/magnum/classMagnum_1_1GL_1_1BufferImage.html), wrapping a PBO, but that's probably not very interoperable with EGL either?

I have to admit I don't have much experience with EGLImages, need to update my knowledge first :)

Reply:
> There's [GL::BufferImage](https://doc.magnum.graphics/magnum/classMagnum_1_1GL_1_1BufferImage.html), wrapping a PBO, but that's probably not very interoperable with EGL either?
> 

Cool. I've never actually used PBOs before but it looks like they allow you to copy GPU memory into the CPU address space in an efficient way.

EGLImage lets you create a reference (not a copy) to an existing GL object (Texture, RenderBuffer) that you can share with another EGL client. You can also share an EGLImage between processes using a platform-specific EGL extension or API (e.g. https://www.khronos.org/registry/EGL/extensions/KHR/EGL_KHR_image_pixmap.txt)

To actually access the EGLImage in CPU memory, you'd need to do a copy.  You'd need to import the EGLImage as a RenderBuffer or Texture and then glReadPixels() or maybe use a PBO to access the data. EGLImage is useful if you want to pass the data by reference to another EGL client (CUDA, another EGL rendering context, etc).

https://www.khronos.org/registry/EGL/extensions/KHR/EGL_KHR_image_base.txt
https://www.khronos.org/registry/OpenGL/extensions/OES/OES_EGL_image.txt
https://docs.nvidia.com/cuda/cuda-for-tegra-appnote/index.html